### PR TITLE
Replace `if ... in spec` with `spec.satisfies` in d* and e* packages

### PR DESCRIPTION
**Pull request by aurianer**
with the structure if "+variant" in spec:, checks like if "mpi" in spec: would be valid whereas it could be ~mpi or +mpi. Satisfies adds additional checks to avoid mistakes from the user.

### CODE REVIEW

File: var/spack/repos/builtin/packages/esmf/package.py
Comment (wdconinc) [inline var/spack/repos/builtin/packages/esmf/package.py]:
```suggestion
            if self.spec.satisfies("^[virtuals=mpi] cray-mpich"):
```

File: var/spack/repos/builtin/packages/esmf/package.py
Comment (wdconinc) [inline var/spack/repos/builtin/packages/esmf/package.py]:
```suggestion
            elif spec.satisfies("^[virtuals=mpi] mvapich2"):
```

File: var/spack/repos/builtin/packages/esmf/package.py
Comment (wdconinc) [inline var/spack/repos/builtin/packages/esmf/package.py]:
```suggestion
            elif spec.satisfies("^[virtuals=mpi] mpich"):
```

File: var/spack/repos/builtin/packages/esmf/package.py
Comment (wdconinc) [inline var/spack/repos/builtin/packages/esmf/package.py]:
```suggestion
            elif spec.satisfies("^[virtuals=mpi] openmpi") or spec.satisfies("^[virtuals=mpi] hpcx-mpi"):
```
(and probably style fix)

File: var/spack/repos/builtin/packages/esmf/package.py
Comment (wdconinc) [inline var/spack/repos/builtin/packages/esmf/package.py]:
```suggestion
            elif spec.satisfies("^[virtuals=mpi] mpt"):
```

File: var/spack/repos/builtin/packages/extrae/package.py
Comment (wdconinc) [inline var/spack/repos/builtin/packages/extrae/package.py]:
```suggestion
        if spec.satisfies("^[virtuals=mpi] intel-oneapi-mpi"):
```

**Review by wdconinc — COMMENTED** [review]
Two packages with mpi virtuals where we will want to be more explicit: `esmf` (@climbfuji et al) and `extrae` (no maintainer).

As in #46127, a better PR description would be great.

**Comment by climbfuji** [comment]
> Two packages with mpi virtuals where we will want to be more explicit: `esmf` (@climbfuji et al) and `extrae` (no maintainer).
> 
> As in #46127, a better PR description would be great.

@theurich @AlexanderRichert-NOAA @jedwards4b @uturuncoglu Please note the suggested change for ESMF in this PR. Thanks @wdconinc for notifying me, no idea why the maintainers of the packages affected by this PR weren't notified automatically.

**Comment by wdconinc** [comment]
> no idea why the maintainers of the packages affected by this PR weren't notified automatically.

Probably just too many :-) There's a cutoff on the number of maintainers that are (can be) notified in a PR.

**Comment by wdconinc** [comment]
Rebasing to pick up CI fix in https://github.com/spack/spack/commit/1c1970e7277440069c0050ebaf6720f72dcdf2e5

**Comment by aurianer** [comment]
@wdconinc this is good to go on my side, let me know if I should change anything still

**Review by wdconinc — DISMISSED** [review]
This looks good to me. Thanks!

Just waiting for @climbfuji to approve as well since `esmf` does not seem to be in any cloud pipelines.

**Comment by climbfuji** [comment]
I would like to test this before it gets merged. Will do today

File: var/spack/repos/builtin/packages/esmf/package.py
Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/esmf/package.py]: It doesn't look like this was incorporated?

**Comment by AlexanderRichert-NOAA** [comment]
Would it make sense to consolidate `if spec.satisfies("x") and spec.satisfies("y")` constructs (e.g., in [expat](https://github.com/spack/spack/blob/56a370a7cde9febdc295f4650b5ab970907319bf/var/spack/repos/builtin/packages/expat/package.py#L179)) into `if spec.satisfies("x y")`? In any case it could always be done later one package at a time, but I thought I'd throw out the idea.

File: var/spack/repos/builtin/packages/esmf/package.py
Comment (wdconinc) [reply var/spack/repos/builtin/packages/esmf/package.py]: Good catch.

**Review by wdconinc — APPROVED** [review]
Thanks @AlexanderRichert-NOAA, now both remaining suggestions applied as well.

**Review by climbfuji — APPROVED** [review]
Installed ESMF successfully on Rocky Linux 9.1 with gcc@11.4.1. Also checked changes for two other packages I care about (ecmwf-atlas, eccodes). Thanks for the updates!

**Pull request merged**
